common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1869 @@
+	attempts := []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		attempts = append(attempts, as.FetchTxAttempts(states, filter)...)
+	}
+	// sort by tx_id ASC, gas_price DESC, gas_tip_cap DESC
+	sort.SliceStable(attempts, func(i, j int) bool {
+		return attempts[i].TxID < attempts[j].TxID
+	})
+
+	// deep copy the attempts
+	var eAttempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	for _, attempt := range attempts {
+		eAttempts = append(eAttempts, ms.deepCopyTxAttempt(attempt.Tx, attempt))
+	}
+
+	return eAttempts, nil
+}
+
+// UpdateBroadcastAts updates the broadcast_at time for a given set of attempts
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateBroadcastAts(ctx context.Context, now time.Time, txIDs []int64) error {
+	// Persist to persistent storage
+	if err := ms.txStore.UpdateBroadcastAts(ctx, now, txIDs); err != nil {
+		return err
+	}
+
+	// Update in memory store
+	fn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+		if tx.BroadcastAt != nil && tx.BroadcastAt.Before(now) {
+			tx.BroadcastAt = &now
+		}
+	}
+
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		as.ApplyToTxsByState(nil, fn, txIDs...)
+	}
+
+	return nil
+}
+
+// UpdateTxsUnconfirmed updates the unconfirmed transactions for a given set of ids
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateTxsUnconfirmed(ctx context.Context, txIDs []int64) error {
+	// Persist to persistent storage
+	if err := ms.txStore.UpdateTxsUnconfirmed(ctx, txIDs); err != nil {
+		return err
+	}
+
+	// Update in memory store
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	wg := sync.WaitGroup{}
+	for _, as := range ms.addressStates {
+		wg.Add(1)
+		go func(as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) {
+			for _, txID := range txIDs {
+				if err := as.MoveConfirmedMissingReceiptToUnconfirmed(txID); err != nil {
+					continue
+				}
+			}
+			wg.Done()
+		}(as)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// FindTxAttemptsRequiringReceiptFetch returns all transactions that are missing a receipt
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxAttemptsRequiringReceiptFetch(ctx context.Context, chainID CHAIN_ID) (
+	attempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	err error,
+) {
+	if ms.chainID.String() != chainID.String() {
+		return attempts, fmt.Errorf("find_tx_attempts_requiring_receipt_fetch: %w", ErrInvalidChainID)
+	}
+
+	filterFn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.TxAttempts != nil && len(tx.TxAttempts) > 0 {
+			attempt := tx.TxAttempts[0]
+			return attempt.State != txmgrtypes.TxAttemptInsufficientFunds
+		}
```

Comment:
Couldn't a tx have attempts with different states? In which case, I think we'd have to traverse the whole attempts list for at least one to match the `TxAttemptInsufficientFunds` state.